src/include/liburing.h

```diff
@@ -1093,20 +1093,21 @@ static inline unsigned io_uring_sq_space_left(const struct io_uring *ring)
 }
 
 /*
- * Only applicable when using SQPOLL - allows the caller to wait for space
- * to free up in the SQ ring, which happens when the kernel side thread has
- * consumed one or more entries. If the SQ ring is currently non-full, no
- * action is taken. Note: may return -EINVAL if the kernel doesn't support
+ * Only applicable when using SQPOLL - allows the caller to wait for an at
+ * least one free entry in the SQ ring, which happens when the kernel side
+ * thread has consumed one or more entries. If the SQ ring is currently
+ * non-full, no waiting occured and an amout of free entries is returned
+ * immediately. Note: may also return -EINVAL if the kernel doesn't support
  * this feature.
  */
 static inline int io_uring_sqring_wait(struct io_uring *ring)
 {
-	if (!(ring->flags & IORING_SETUP_SQPOLL))
-		return 0;
-	if (io_uring_sq_space_left(ring))
-		return 0;
+	int ret;
 
-	return __io_uring_sqring_wait(ring);
+	if (!(ring->flags & IORING_SETUP_SQPOLL))
+		return -EINVAL;
+	return (ret = io_uring_sq_space_left(ring))
```

Comment:
This is unreadable, please write this without using a ternary. It's made worse by the fact that you double the assignment as the value as well.

And if re-written a bit, it also means you don't need a kernel change.

Reply:
__it also means you don't need a kernel change__

Are you sure? I didn't dig the kernel part too much but it seems that `io_uring_enter(fd, 0, 0,  IORING_ENTER_SQ_WAIT, NULL)` will return 0 without this change.

Reply:
If the kernel side returns 0, you'd use the helper to return available entries. You don't need the kernel to return this for you.

Reply:
OK but this requires an extra call to helper (which hopefully may be avoided with just one-line kernel change).